File: fetch_asset.php
<?php
if(count($argv) < 2){
	echo "usage $argv[0] ip1 ip2 ... \n";
	exit;
}
function isIp($ip){
	if(!preg_match("/^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/", $ip)) {
		echo "$ip is not a valid ip address!\n";
		return false;
	}
	return true;
}
function get($url){
	    $ch = curl_init();
		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
		$content = curl_exec($ch);
		curl_close($ch);
		return $content;
}
function getProcInfo($serverId){
	
	$url='http://cmdb1.sysop.duowan.com:8088/webservice/serverProc/getByServerId.do?server_id='.$serverId;
	$content=get($url);	
	return $content;
}
function formatHostIp($ipstr){
	$ip_array=explode('-',$ipstr);	
	switch($ip_array[0]){
		case 1:	$hostinfo="双线 ".$ip_array[1];break;
		case 2:	$hostinfo="电信 ".$ip_array[1];break;
		case 3:	$hostinfo="联通 ".$ip_array[1];break;
		case 4:	$hostinfo="电信 ".$ip_array[1];break;
		case 5:	$hostinfo="联通 ".$ip_array[1];break;
		case 6:	$hostinfo="BGP ".$ip_array[1];break;
		case 7:	$hostinfo="教育网 ".$ip_array[1];break;
		case 8:	$hostinfo="移动 ".$ip_array[1];break;
		case 9:	$hostinfo="内网 ".$ip_array[1];break;
		case 14:	$hostinfo="香港 ".$ip_array[1];break;
		case 16:	$hostinfo="巴西 ".$ip_array[1];break;
	}
	return $hostinfo;
}
function connectMemcache(){
	$mem_ip='183.61.143.222';
	$mem_port='11211';
	$mem = new Memcache;
	if(!$mem->connect($mem_ip,$mem_port)){
		echo "memcache can not connect!";
		exit;
	}
	else{
		return $mem;
	}
}
function closeMemConnect($mem){
	if(!$mem->close()){
		echo 'cannot close memcache connect!';
	}
}
function getHost($ip){
	$mem=connectMemcache();
	$guid=$mem->get($ip);
	if($guid){
		$host=$mem->get($guid);
		return $host;
	}
	else{
		echo "no ip info!";
		return false;
	}
	closeMemConnect($mem);
}
function printHost($host){
	$hostinfo="";
	foreach ($host['ips'] as $ipstr ){
		$hostinfo.=formatHostIp($ipstr)." | ";	
	}
	$procinfo=getProcInfo($host['serverId']);
	switch($host['status']){
		case 1:	$status="运营中";break;
		case 2:	$status="报修";break;
		case 3:	$status="迁移中";break;
		case 4:	$status="闲置";break;
		case 6:	$status="库存";break;
		case 7:	$status="测试中";break;
		case 8:	$status="开发使用中";break;
		case 9:	$status="报废";break;
		default:	$status="其他";break;
	}
	echo "-----------------------------------------------\n";
	echo "|服务器IP:	|	".$hostinfo."\n";
	echo "-----------------------------------------------\n";
	if(!empty($host['internal_ip'])){
	echo "|内网IP:	|	".implode(',',$host['internal_ip'])."\n";
	echo "-----------------------------------------------\n";}
	echo "|服务器ID:	|	".$host['serverId']."\n";
	echo "-----------------------------------------------\n";
	echo "|机房GroupID:	|	".$host['priGroupId']."\n";
	echo "-----------------------------------------------\n";
	echo "|机房:		|	".$host['roomName']."\n";
	echo "-----------------------------------------------\n";
	echo "|状态:		|	".$status."\n";
	echo "-----------------------------------------------\n";
	echo "|业务模块:	|	".$host['buss']."\n";
	echo "-----------------------------------------------\n";
	echo "|开发负责人:	|	".$host['responsibleAdmin']."\n";
	echo "-----------------------------------------------\n";
	echo "|运维负责人:	|	".$host['sysopResponsibleAdmin']."\n";
	echo "-----------------------------------------------\n";
	echo "	".$host['ip']." 进程列表:		\n";
	echo "***********************************************\n";
	echo preg_replace('/ /',' * ',$procinfo);
	echo "***********************************************\n";
}

array_shift($argv);
foreach($argv as $ip){
	if(isIp($ip)){
		$host=getHost($ip);
		$host && printHost($host);
	}
	else{
		continue;
	}
}
?>

File: fetch_asset.py
#!/usr/bin/python
#coding:utf-8

import pymongo
import urllib
import sys
reload(sys)
sys.setdefaultencoding('utf-8') 

def connect_db(dbName,collect):
	dbInfo = [ '127.0.0.1' , '27017']
	conn = pymongo.Connection(dbInfo)
	return conn[dbName][collect]

def get_serverId(ip):
	collection = connect_db("hosts","hosts")
	idInfo = collection.find_one({ "_id" : ip })
	if not idInfo:
		print ip + " not exist in cmdb!"
		sys.exit()
	serverId = str(idInfo['serverId'])
	return serverId

def get_hostinfo(ip):
	serverId = get_serverId(ip)
	collection = connect_db("hosts","hosts")
	hostInfo = collection.find_one({ "_id" : serverId })
	return hostInfo

def print_split_line(symbol = "-"):
	print symbol * 60

def print_attribute(key,value):
	print_split_line()
	print "|\t" + str(key) + "\t | \t" + str(value)

def get_proc_info(ip):
	serverId = get_serverId(ip)
	url = 'http://cmdb1.sysop.duowan.com:8088/webservice/serverProc/getByServerId.do?server_id=' + serverId
	content = urllib.urlopen(url).read()
	return content


def print_host_stats(ip):
	hostInfo = get_hostinfo(ip)
	status_map = { '1' : '运营中' , '2' : '报修中' , '3' : '迁移中' , '4' : '库存' , '7' : '测试中' , '9' : '报废' }
	isp_map = { '4':'电信','5':'联通','6':'BGP','7':'教育','8':'移动','14':'香港','16':'巴西' }

	ipstr = ""
	for isp_code in hostInfo['ips']:
		ipstr += isp_map.get(isp_code,'error') + " : " + hostInfo['ips'][isp_code] + " "
	print_attribute("服务器IP",ipstr)

	if "internal_ip" in hostInfo:
		internal_ipstr = ""
		for inter_ip in hostInfo['internal_ip']:
			internal_ipstr += inter_ip + " " 
		print_attribute("内网  IP",internal_ipstr)
	
	print_attribute("ServerId",hostInfo['serverId'])
	print_attribute("服务器状态",status_map.get(str(hostInfo['status']),'error'))
	print_attribute("机房GroupId",hostInfo['priGroupId'])
	print_attribute("机房名称",hostInfo['roomName'])
	print_attribute("业务模块",hostInfo['buss'])
	print_attribute("开发负责人",hostInfo['responsibleAdmin'])
	print_attribute("运维负责人",hostInfo['sysopResponsibleAdmin'])
	print_split_line("*")
	print "进程列表".center(60)
	print_split_line("*")
	procInfo = get_proc_info(ip)
	print procInfo
	print_split_line("*")

ip = sys.argv[1]
print_host_stats(ip)
